Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs

```diff
@@ -0,0 +1,181 @@
+    /// are specific to Azure App Services (Web Apps, Functions, etc...).
+    /// </summary>
+    public sealed class AppServicesHeartbeatTelemetryModule : ITelemetryModule, IDisposable
+    {
+        /// <summary>
+        /// Environment variables and the Application Insights heartbeat field names that accompany them.
+        /// </summary>
+        internal static readonly KeyValuePair<string, string>[] WebHeartbeatPropertyNameEnvVarMap = new KeyValuePair<string, string>[]
+        {
+            new KeyValuePair<string, string>("appSrv_SiteName", "WEBSITE_SITE_NAME"),
+            new KeyValuePair<string, string>("appSrv_wsStamp", "WEBSITE_HOME_STAMPNAME"),
+            new KeyValuePair<string, string>("appSrv_wsHost", "WEBSITE_HOSTNAME")
+        };
+
+        // for testing only: override the heartbeat manager
+        internal IHeartbeatPropertyManager HeartbeatManager;
+
+        // to provide a 'singleton' accessor for updating the env vars should they change during runtime
+        private static AppServicesHeartbeatTelemetryModule instance;
+        private bool isInitialized = false;
+        
+        /// <summary>
+        /// Initializes a new instance of the<see cref="AppServicesHeartbeatTelemetryModule" /> class.
+        /// </summary>
+        public AppServicesHeartbeatTelemetryModule() : this(null)
+        {            
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AppServicesHeartbeatTelemetryModule" /> class. This is
+        /// internal, and allows for overriding the Heartbeat Property Manager to test this module with.
+        /// </summary>
+        /// <param name="hbeatPropManager">The heartbeat property manager to use when setting/updating env var values.</param>
+        internal AppServicesHeartbeatTelemetryModule(IHeartbeatPropertyManager hbeatPropManager)
+        {
+            this.HeartbeatManager = hbeatPropManager;
+            AppServicesHeartbeatTelemetryModule.Instance = this;
+        }
+
+        /// <summary>
+        /// Gets a value to provide internal access to the only instance of this class that *should* be available.
+        /// </summary>
+        public static AppServicesHeartbeatTelemetryModule Instance
+        {
+            get => AppServicesHeartbeatTelemetryModule.instance;
+            private set => AppServicesHeartbeatTelemetryModule.instance = value;
+        }
+
+        /// <summary>
+        /// Initialize the default heartbeat provider for Azure App Services. This module
+        /// looks for specific environment variables and sets them into the heartbeat 
+        /// properties for Application Insights, if they exist.
+        /// </summary>
+        /// <param name="configuration">Unused parameter.</param>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            this.isInitialized = this.UpdateHeartbeatWithAppServiceEnvVarValues();
```

Comment:
this.isInitialized -- Where is this used?

Reply:
Line 94 - in the call to AddAppServiceEnvironmentVariablesToHeartbeat. The first time (isInitialized=false) we use AddHeartbeatProperty, each subsequent time we use SetHeartbeatProperty.